### ISSUE TITLE
deadlock on Session clean up

### ISSUE THREAD

**Issue by tpokorra** [openpetra/openpetra#452]
we get this on https://lbs.solidcharity.com/package/solidcharity/openpetra/openpetra-test-setup#test_ubuntu/bionic/amd64

```
Saturday, 14-Mar-2020, 07:16:26.72 : Error executing non-query SQL statement.
on Thread 'Thread Pool Worker' [ThreadID: 6] in AppDomain 'bd2c0138'
The SQL Statement was:
DELETE FROM s_session WHERE s_valid_until_d < NOW()
Possible cause: MySql.Data.MySqlClient.MySqlException (0x80004005): Deadlock found when trying to get lock; try restarting transaction
  at MySql.Data.MySqlClient.MySqlStream.ReadPacket () [0x00096] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.NativeDriver.GetResult (System.Int32& affectedRow, System.Int64& insertedId) [0x00026] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.Driver.GetResult (System.Int32 statementId, System.Int32& affectedRows, System.Int64& insertedId) [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.Driver.NextResult (System.Int32 statementId, System.Boolean force) [0x00023] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlDataReader.NextResult () [0x00205] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0030c] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader () [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery () [0x00032] in <091acd56583e48c19623fa46d3a21a2a>:0
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x000a2] in <6d939accbca64f6d9a704e907d217e23>:0
Message: Deadlock found when trying to get lock; try restarting transaction
MySQL Error Number: 1213

Saturday, 14-Mar-2020, 07:16:26.73 :     at Void LogStackTrace(Ict.Common.TLoggingType)
    at Void LogException(System.Exception, System.String, System.Data.Common.DbParameter[], System.String, Boolean)
    at Void LogExceptionAndThrow(System.Exception, System.String, System.Data.Common.DbParameter[], System.String)
    at Int32 ExecuteNonQuery(System.String, Ict.Common.DB.TDBTransaction, Boolean, System.Data.Common.DbParameter[], Boolean)
    at Int32 ExecuteNonQuery(System.String, Ict.Common.DB.TDBTransaction, System.Data.Common.DbParameter[], Boolean)
    at Void CleanOldSessions(Ict.Common.DB.TDBTransaction)
    at Void InitSession(System.String, Ict.Common.DB.TDBTransaction)
    at Void <>m__0()
    at Void TransactionInner(Ict.Common.DB.TDBTransaction ByRef, System.Data.IsolationLevel, Boolean ByRef, System.Action, System.String)
    at Void WriteTransaction(Ict.Common.DB.TDBTransaction ByRef, Boolean ByRef, System.Action)
    at Void InitThread(System.String)
    at Boolean Init()
    at Void .ctor()
    at System.Object InternalInvoke(System.Reflection.MonoCMethod, System.Object, System.Object[], System.Exception ByRef)
    at System.Object InternalInvoke(System.Object, System.Object[], Boolean)
    at System.Object CreateInstanceMono(Boolean, Boolean)
    at System.Object CreateInstanceSlow(Boolean, Boolean, Boolean, Boolean, System.Threading.StackCrawlMark ByRef)
    at System.Object CreateInstanceDefaultCtor(Boolean, Boolean, Boolean, Boolean, System.Threading.StackCrawlMark ByRef)
    at System.Object CreateInstance(System.Type, Boolean, Boolean)
    at System.Object CreateInstance(System.Type, Boolean)
    at System.Object CreateInstance(System.Type)
    at Void CreateServerInstance()
....
```

**Comment by tpokorra** [comment]
still getting:

```
Thursday, 19-Mar-2020, 04:50:03.84 : Error executing non-query SQL statement.
on Thread 'Thread Pool Worker' [ThreadID: 6] in AppDomain 'bd2c0138'
The SQL Statement was:
DELETE FROM s_session WHERE s_valid_until_d < NOW()
Possible cause: MySql.Data.MySqlClient.MySqlException (0x80004005): Deadlock found when trying to get lock; try restarting transaction
  at MySql.Data.MySqlClient.MySqlStream.ReadPacket () [0x00096] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.NativeDriver.GetResult (System.Int32& affectedRow, System.Int64& insertedId) [0x00026] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.Driver.GetResult (System.Int32 statementId, System.Int32& affectedRows, System.Int64& insertedId) [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.Driver.NextResult (System.Int32 statementId, System.Boolean force) [0x00023] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlDataReader.NextResult () [0x00205] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0030c] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader () [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery () [0x00032] in <091acd56583e48c19623fa46d3a21a2a>:0
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x000a2] in <6d939accbca64f6d9a704e907d217e23>:0
Message: Deadlock found when trying to get lock; try restarting transaction
MySQL Error Number: 1213

Thursday, 19-Mar-2020, 04:50:03.85 :     at Void LogStackTrace(Ict.Common.TLoggingType)
    at Void LogException(System.Exception, System.String, System.Data.Common.DbParameter[], System.String, Boolean)
    at Void LogExceptionAndThrow(System.Exception, System.String, System.Data.Common.DbParameter[], System.String)
    at Int32 ExecuteNonQuery(System.String, Ict.Common.DB.TDBTransaction, Boolean, System.Data.Common.DbParameter[], Boolean)
    at Int32 ExecuteNonQuery(System.String, Ict.Common.DB.TDBTransaction, System.Data.Common.DbParameter[], Boolean)
    at Void CleanOldSessions(Ict.Common.DB.TDBTransaction)
    at Void InitSession(System.String, Ict.Common.DB.TDBTransaction)
    at Void <>m__0()
    at Void TransactionInner(Ict.Common.DB.TDBTransaction ByRef, System.Data.IsolationLevel, Boolean ByRef, System.Action, System.String)
    at Void WriteTransaction(Ict.Common.DB.TDBTransaction ByRef, Boolean ByRef, System.Action)
    at Void InitThread(System.String)
    at Boolean Init()
    at Void .ctor()
[...]

Thursday, 19-Mar-2020, 04:50:03.87 : Exception in TOpenPetraOrgSessionManager.Init()

Thursday, 19-Mar-2020, 04:50:03.87 : Ict.Common.Exceptions.EOPDBException: [Context: Error executing non-query SQL statement.] ---> MySql.Data.MySqlClient.MySqlException: Deadlock found when trying to get lock; try restarting transaction
  at MySql.Data.MySqlClient.MySqlStream.ReadPacket () [0x00096] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.NativeDriver.GetResult (System.Int32& affectedRow, System.Int64& insertedId) [0x00026] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.Driver.GetResult (System.Int32 statementId, System.Int32& affectedRows, System.Int64& insertedId) [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.Driver.NextResult (System.Int32 statementId, System.Boolean force) [0x00023] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlDataReader.NextResult () [0x00205] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0030c] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader () [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery () [0x00032] in <091acd56583e48c19623fa46d3a21a2a>:0
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x000a2] in <6d939accbca64f6d9a704e907d217e23>:0
   --- End of inner exception stack trace ---
  at Ict.Common.DB.TDataBase.LogException (System.Exception AException, System.String ASqlStatement, System.Data.Common.DbParameter[] AParametersArray, System.String AContext, System.Boolean AThrowExceptionAfterLogging) [0x003ab] in <6d939accbca64f6d9a704e907d217e23>:0
  at Ict.Common.DB.TDataBase.LogExceptionAndThrow (System.Exception AException, System.String ASqlStatement, System.Data.Common.DbParameter[] AParametersArray, System.String AContext) [0x00000] in <6d939accbca64f6d9a704e907d217e23>:0
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x0012f] in <6d939accbca64f6d9a704e907d217e23>:0
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x00000] in <6d939accbca64f6d9a704e907d217e23>:0
  at Ict.Common.Session.TSession.CleanOldSessions (Ict.Common.DB.TDBTransaction AWriteTransaction) [0x0000c] in <e5e2f192c90a49939f0f34aa58ac0868>:0
  at Ict.Common.Session.TSession.InitSession (System.String ASessionID, Ict.Common.DB.TDBTransaction AWriteTransaction) [0x000d5] in <e5e2f192c90a49939f0f34aa58ac0868>:0
  at Ict.Common.Session.TSession+<InitThread>c__AnonStorey0.<>m__0 () [0x00000] in <e5e2f192c90a49939f0f34aa58ac0868>:0
  at Ict.Common.DB.TDataBase.TransactionInner (Ict.Common.DB.TDBTransaction& ATransaction, System.Data.IsolationLevel AIsolationLevel, System.Boolean& ASubmitOK, System.Action AEncapsulatedDBAccessCode, System.String AContext) [0x000f7] in <6d939accbca64f6d9a704e907d217e23>:0
  at Ict.Common.DB.TDataBase.WriteTransaction (Ict.Common.DB.TDBTransaction& ATransaction, System.Boolean& ASubmitOK, System.Action AEncapsulatedDBAccessCode) [0x00000] in <6d939accbca64f6d9a704e907d217e23>:0
  at Ict.Common.Session.TSession.InitThread (System.String ASessionID) [0x00074] in <e5e2f192c90a49939f0f34aa58ac0868>:0
  at Ict.Petra.Server.App.WebService.TOpenPetraOrgSessionManager.Init () [0x00173] in <36a8dcea213a4bbfb8dcc68d9d479b5f>:0
  at Ict.Petra.Server.App.WebService.TOpenPetraOrgSessionManager..ctor () [0x00006] in <36a8dcea213a4bbfb8dcc68d9d479b5f>:0



```

**Comment by tpokorra** [comment]
ah, that was on test, and the fix is not in the latest openpetra bin yet...

**Comment by tpokorra** [comment]
still happens sometimes:

```
Friday, 20-Mar-2020, 07:09:40.43 : Error executing non-query SQL statement.
on Thread 'Thread Pool Worker' [ThreadID: 6] in AppDomain 'bd2c0138'
The SQL Statement was: 
INSERT INTO s_session (s_session_values_c, s_session_id_c, s_valid_until_d) VALUES (?,?,?)
Parameter: 1 s_session_values_c {} System.String Text 2
Parameter: 2 s_session_id_c 350c9967-9f34-4190-9057-b15dfa70668c System.String VarChar 36
Parameter: 3 s_valid_until_d 3/21/2020 7:09:40 AM System.DateTime DateTime 0
Possible cause: MySql.Data.MySqlClient.MySqlException (0x80004005): Deadlock found when trying to get lock; try restarting transaction
  at MySql.Data.MySqlClient.MySqlStream.ReadPacket () [0x00096] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.NativeDriver.GetResult (System.Int32& affectedRow, System.Int64& insertedId) [0x00026] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.Driver.GetResult (System.Int32 statementId, System.Int32& affectedRows, System.Int64& insertedId) [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.Driver.NextResult (System.Int32 statementId, System.Boolean force) [0x00023] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlDataReader.NextResult () [0x00205] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0030c] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader () [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery () [0x00032] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x000a2] in <daeda98150e44282a91cb91fb6f5a690>:0 
Message: Deadlock found when trying to get lock; try restarting transaction
MySQL Error Number: 1213


Friday, 20-Mar-2020, 07:09:40.44 :     at Void LogStackTrace(Ict.Common.TLoggingType)
    at Void LogException(System.Exception, System.String, System.Data.Common.DbParameter[], System.String, Boolean)
    at Void LogExceptionAndThrow(System.Exception, System.String, System.Data.Common.DbParameter[], System.String)
    at Int32 ExecuteNonQuery(System.String, Ict.Common.DB.TDBTransaction, Boolean, System.Data.Common.DbParameter[], Boolean)
    at Int32 ExecuteNonQuery(System.String, Ict.Common.DB.TDBTransaction, System.Data.Common.DbParameter[], Boolean)
    at Void SaveSession(Ict.Common.DB.TDBTransaction)
    at Boolean InitSession(System.String, Ict.Common.DB.TDBTransaction)
    at Void <>m__0()
    at Void TransactionInner(Ict.Common.DB.TDBTransaction ByRef, System.Data.IsolationLevel, Boolean ByRef, System.Action, System.String)
    at Void WriteTransaction(Ict.Common.DB.TDBTransaction ByRef, Boolean ByRef, System.Action)
    at Void InitThread(System.String)
    at Boolean Init()
    at Void .ctor()
    at System.Object InternalInvoke(System.Reflection.MonoCMethod, System.Object, System.Object[], System.Exception ByRef)
    at System.Object InternalInvoke(System.Object, System.Object[], Boolean)
    at System.Object CreateInstanceMono(Boolean, Boolean)
    at System.Object CreateInstanceSlow(Boolean, Boolean, Boolean, Boolean, System.Threading.StackCrawlMark ByRef)
    at System.Object CreateInstanceDefaultCtor(Boolean, Boolean, Boolean, Boolean, System.Threading.StackCrawlMark ByRef)
    at System.Object CreateInstance(System.Type, Boolean, Boolean)
    at System.Object CreateInstance(System.Type, Boolean)
    at System.Object CreateInstance(System.Type)
    at Void CreateServerInstance()
    at Void Invoke()
    at Void CoreProcessRequest()
    at Void ProcessRequest(System.Web.HttpContext)
    at Boolean MoveNext()
    at Void Tick()
    at Void Start(System.Object)
    at Void System.Web.IHttpHandler.ProcessRequest(System.Web.HttpContext)
    at Void Process(System.Web.HttpWorkerRequest)
    at Void RealProcessRequest(System.Object)
    at Void ProcessRequest(System.Web.HttpWorkerRequest)
    at Void ProcessRequest()
    at Void ProcessRequest(Mono.WebServer.MonoWorkerRequest)
    at Void ProcessRequest(Mono.WebServer.FastCgi.Responder)
    at Int32 Process()
    at Void Worker(System.Object)
    at Void OnInputDataReceived(Mono.FastCgi.Request, Mono.FastCgi.DataReceivedArgs)
    at Void AddInputData(Mono.WebServer.FastCgi.Record)
    at Void HandleStandardInput(Mono.FastCgi.Request, Mono.WebServer.FastCgi.Record)
    at Void HandleRequest(Mono.WebServer.FastCgi.Record, Mono.FastCgi.Request)
    at Void Run()
    at Void Run()
    at Void OnAccept(System.IAsyncResult)
    at Void <Complete>m__0(System.Object)
    at Void System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
    at Boolean Dispatch()
    at Boolean PerformWaitCallback()
in Appdomain bd2c0138


Friday, 20-Mar-2020, 07:09:40.44 : Exception in TOpenPetraOrgSessionManager.Init()

Friday, 20-Mar-2020, 07:09:40.45 : Ict.Common.Exceptions.EOPDBException: [Context: Error executing non-query SQL statement.] ---> MySql.Data.MySqlClient.MySqlException: Deadlock found when trying to get lock; try restarting transaction
  at MySql.Data.MySqlClient.MySqlStream.ReadPacket () [0x00096] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.NativeDriver.GetResult (System.Int32& affectedRow, System.Int64& insertedId) [0x00026] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.Driver.GetResult (System.Int32 statementId, System.Int32& affectedRows, System.Int64& insertedId) [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.Driver.NextResult (System.Int32 statementId, System.Boolean force) [0x00023] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlDataReader.NextResult () [0x00205] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0030c] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader () [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery () [0x00032] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x000a2] in <daeda98150e44282a91cb91fb6f5a690>:0 
   --- End of inner exception stack trace ---
  at Ict.Common.DB.TDataBase.LogException (System.Exception AException, System.String ASqlStatement, System.Data.Common.DbParameter[] AParametersArray, System.String AContext, System.Boolean AThrowExceptionAfterLogging) [0x003ab] in <daeda98150e44282a91cb91fb6f5a690>:0 
  at Ict.Common.DB.TDataBase.LogExceptionAndThrow (System.Exception AException, System.String ASqlStatement, System.Data.Common.DbParameter[] AParametersArray, System.String AContext) [0x00000] in <daeda98150e44282a91cb91fb6f5a690>:0 
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x0012f] in <daeda98150e44282a91cb91fb6f5a690>:0 
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x00000] in <daeda98150e44282a91cb91fb6f5a690>:0 
  at Ict.Common.Session.TSession.SaveSession (Ict.Common.DB.TDBTransaction AWriteTransaction) [0x00112] in <a239ad1ffc2d489795c9ce4c3c27a228>:0 
  at Ict.Common.Session.TSession.InitSession (System.String ASessionID, Ict.Common.DB.TDBTransaction AWriteTransaction) [0x000e7] in <a239ad1ffc2d489795c9ce4c3c27a228>:0 
  at Ict.Common.Session.TSession+<InitThread>c__AnonStorey0.<>m__0 () [0x00000] in <a239ad1ffc2d489795c9ce4c3c27a228>:0 
  at Ict.Common.DB.TDataBase.TransactionInner (Ict.Common.DB.TDBTransaction& ATransaction, System.Data.IsolationLevel AIsolationLevel, System.Boolean& ASubmitOK, System.Action AEncapsulatedDBAccessCode, System.String AContext) [0x000f7] in <daeda98150e44282a91cb91fb6f5a690>:0 
  at Ict.Common.DB.TDataBase.WriteTransaction (Ict.Common.DB.TDBTransaction& ATransaction, System.Boolean& ASubmitOK, System.Action AEncapsulatedDBAccessCode) [0x00000] in <daeda98150e44282a91cb91fb6f5a690>:0 
  at Ict.Common.Session.TSession.InitThread (System.String ASessionID) [0x0007b] in <a239ad1ffc2d489795c9ce4c3c27a228>:0 
  at Ict.Petra.Server.App.WebService.TOpenPetraOrgSessionManager.Init () [0x00173] in <5671faf71cbc472b8c151ba16eca6753>:0 
  at Ict.Petra.Server.App.WebService.TOpenPetraOrgSessionManager..ctor () [0x00006] in <5671faf71cbc472b8c151ba16eca6753>:0 
```

**Comment by tpokorra** [comment]
trying to reproduce:

```
OP_CUSTOMER=op_test /home/openpetra/openpetra-server.sh initdb
systemctl restart openpetra
systemctl restart nginx
sleep 5
curl --silent --location http://localhost
curl --silent --location http://localhost/api/serverSessionManager.asmx/IsUserLoggedIn?
```

**Comment by tpokorra** [comment]
I get now duplicate keys? Guid is unique. So something running in parallel?

```
aturday, 21-Mar-2020, 08:11:37.49 : Error executing non-query SQL statement.
on Thread 'Thread Pool Worker' [ThreadID: 6] in AppDomain 'bd2c0138'
The SQL Statement was:
INSERT INTO s_session (s_session_values_c, s_session_id_c, s_valid_until_d) VALUES (?,?,?)
Parameter: 1 s_session_values_c {} System.String Text 2
Parameter: 2 s_session_id_c 180340da-298d-444f-b0eb-2c2ee9412a8a System.String VarChar 36
Parameter: 3 s_valid_until_d 3/22/2020 8:11:37 AM System.DateTime DateTime 0
Possible cause: MySql.Data.MySqlClient.MySqlException (0x80004005): Duplicate entry '180340da-298d-444f-b0eb-2c2ee9412a8a' for key 'PRIMARY'
  at MySql.Data.MySqlClient.MySqlStream.ReadPacket () [0x00096] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.NativeDriver.GetResult (System.Int32& affectedRow, System.Int64& insertedId) [0x00026] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.Driver.GetResult (System.Int32 statementId, System.Int32& affectedRows, System.Int64& insertedId) [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.Driver.NextResult (System.Int32 statementId, System.Boolean force) [0x00023] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlDataReader.NextResult () [0x00205] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0030c] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader () [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery () [0x00032] in <091acd56583e48c19623fa46d3a21a2a>:0
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x000a2] in <daeda98150e44282a91cb91fb6f5a690>:0
Message: Duplicate entry '180340da-298d-444f-b0eb-2c2ee9412a8a' for key 'PRIMARY'
MySQL Error Number: 1062

Saturday, 21-Mar-2020, 08:11:37.50 :     at Void LogStackTrace(Ict.Common.TLoggingType)
    at Void LogException(System.Exception, System.String, System.Data.Common.DbParameter[], System.String, Boolean)
    at Void LogExceptionAndThrow(System.Exception, System.String, System.Data.Common.DbParameter[], System.String)
    at Int32 ExecuteNonQuery(System.String, Ict.Common.DB.TDBTransaction, Boolean, System.Data.Common.DbParameter[], Boolean)
    at Int32 ExecuteNonQuery(System.String, Ict.Common.DB.TDBTransaction, System.Data.Common.DbParameter[], Boolean)
    at Void SaveSession(Ict.Common.DB.TDBTransaction)
    at Boolean InitSession(System.String, Ict.Common.DB.TDBTransaction)
    at Void <>m__0()
    at Void TransactionInner(Ict.Common.DB.TDBTransaction ByRef, System.Data.IsolationLevel, Boolean ByRef, System.Action, System.String)
    at Void WriteTransaction(Ict.Common.DB.TDBTransaction ByRef, Boolean ByRef, System.Action)
    at Void InitThread(System.String)
    at Boolean Init()
    at Void .ctor()
```

**Comment by tpokorra** [comment]
see also https://stackoverflow.com/questions/2332768/how-to-avoid-mysql-deadlock-found-when-trying-to-get-lock-try-restarting-trans

**Comment by tpokorra** [comment]
this is not good: the FSessionID is shared between threads? that leads to primary key violation.

```
Saturday, 21-Mar-2020, 09:22:31.17   : TSession: Creating new session: cdd2cd44-5c0e-473d-85f1-b5cbf84a7d50 in Thread 7

Saturday, 21-Mar-2020, 09:22:31.17   : TSession: Creating new session: aa8e839f-8cfe-4cc2-b539-3974bbbfcb47 in Thread 21

Saturday, 21-Mar-2020, 09:22:31.18   : Waiting to obtain Thread-safe access to the Database Abstraction Layer... (Call performed in Thread 'Thread Pool Worker' [ThreadID: 21] in AppDomain '4d01e2ae')

Saturday, 21-Mar-2020, 09:22:31.18   : Waiting to obtain Thread-safe access to the Database Abstraction Layer... (Call performed in Thread 'Thread Pool Worker' [ThreadID: 7] in AppDomain '4d01e2ae')

Saturday, 21-Mar-2020, 09:22:31.18   : Obtained Thread-safe access to the Database Abstraction Layer... (Call performed in Thread 'Thread Pool Worker' [ThreadID: 7] in AppDomain '4d01e2ae')

Saturday, 21-Mar-2020, 09:22:31.18   : Obtained Thread-safe access to the Database Abstraction Layer... (Call performed in Thread 'Thread Pool Worker' [ThreadID: 21] in AppDomain '4d01e2ae')

Saturday, 21-Mar-2020, 09:22:31.18   : (Context: 'Ict.Common.DB.TDataBase.ExecuteScalar() Thread: 7')
The SQL query is: 
SELECT COUNT(*)
FROM PUB_s_session
WHERE s_session_id_c = ?

Saturday, 21-Mar-2020, 09:22:31.18   : (Context: 'Ict.Common.DB.TDataBase.ExecuteScalar() Thread: 21')
The SQL query is: 
SELECT COUNT(*)
FROM PUB_s_session
WHERE s_session_id_c = ?

Saturday, 21-Mar-2020, 09:22:31.18   : Parameter: 1 aa8e839f-8cfe-4cc2-b539-3974bbbfcb47 System.String VarChar 36

Saturday, 21-Mar-2020, 09:22:31.18   : Ict.Common.DB.TDataBase.ExecuteScalar: now creating Command(SELECT COUNT(*) FROM PUB_s_session WHERE s_session_id_c = ?)...

Saturday, 21-Mar-2020, 09:22:31.18   : Parameter: 1 aa8e839f-8cfe-4cc2-b539-3974bbbfcb47 System.String VarChar 36

Saturday, 21-Mar-2020, 09:22:31.18   : Ict.Common.DB.TDataBase.ExecuteScalar: now creating Command(SELECT COUNT(*) FROM PUB_s_session WHERE s_session_id_c = ?)...
```

**Comment by tpokorra** [comment]
we need to use ThreadStatic for such variables: see https://www.c-sharpcorner.com/article/overview-of-threadstatic-attribute-in-c-sharp/

**Comment by tpokorra** [comment]
creating new issue for the ThreadStatic issue: #453

**Comment by tpokorra** [comment]
it is still an issue:

```
Wednesday, 25-Mar-2020, 06:46:53.06 : Error executing non-query SQL statement.
on Thread 'Thread Pool Worker' [ThreadID: 20] in AppDomain 'bd2c0138'
The SQL Statement was: 
INSERT INTO s_session (s_session_values_c, s_session_id_c, s_valid_until_d) VALUES (?,?,?)
Parameter: 1 s_session_values_c {} System.String Text 2
Parameter: 2 s_session_id_c 9930d47c-26c8-402b-ae80-81efb5847b75 System.String VarChar 36
Parameter: 3 s_valid_until_d 3/26/2020 6:46:52 AM System.DateTime DateTime 0
Possible cause: MySql.Data.MySqlClient.MySqlException (0x80004005): Deadlock found when trying to get lock; try restarting transaction
  at MySql.Data.MySqlClient.MySqlStream.ReadPacket () [0x00096] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.NativeDriver.GetResult (System.Int32& affectedRow, System.Int64& insertedId) [0x00026] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.Driver.GetResult (System.Int32 statementId, System.Int32& affectedRows, System.Int64& insertedId) [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.Driver.NextResult (System.Int32 statementId, System.Boolean force) [0x00023] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlDataReader.NextResult () [0x00205] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0030c] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteReader () [0x00000] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery () [0x00032] in <091acd56583e48c19623fa46d3a21a2a>:0 
  at Ict.Common.DB.TDataBase.ExecuteNonQuery (System.String ASqlStatement, Ict.Common.DB.TDBTransaction ATransaction, System.Boolean AMustCoordinateDBAccess, System.Data.Common.DbParameter[] AParametersArray, System.Boolean ACommitTransaction) [0x000a2] in <6a7a71611ccf4f8aa97c0db65250f8ac>:0 
Message: Deadlock found when trying to get lock; try restarting transaction
MySQL Error Number: 1213
```
